src/indexing.jl

```diff
@@ -1,27 +1,85 @@
-argdims(::IndexStyle, ::Type{T}) where {T} = 0
-argdims(::IndexStyle, ::Type{T}) where {T<:Colon} = 1
-argdims(::IndexStyle, ::Type{T}) where {T<:AbstractArray} = ndims(T)
-argdims(::IndexStyle, ::Type{T}) where {N,T<:CartesianIndex{N}} = N
-argdims(::IndexStyle, ::Type{T}) where {N,T<:AbstractArray{CartesianIndex{N}}} = N
-argdims(::IndexStyle, ::Type{T}) where {N,T<:AbstractArray{<:Any,N}} = N
-argdims(::IndexStyle, ::Type{T}) where {N,T<:LogicalIndex{<:Any,<:AbstractArray{Bool,N}}} = N
-@generated function argdims(s::IndexStyle, ::Type{T}) where {N,T<:Tuple{Vararg{<:Any,N}}}
+argdims(::ArrayStyle, ::Type{T}) where {T} = 0
+argdims(::ArrayStyle, ::Type{T}) where {T<:Colon} = 1
+argdims(::ArrayStyle, ::Type{T}) where {T<:AbstractArray} = ndims(T)
+argdims(::ArrayStyle, ::Type{T}) where {N,T<:CartesianIndex{N}} = N
+argdims(::ArrayStyle, ::Type{T}) where {N,T<:AbstractArray{CartesianIndex{N}}} = N
+argdims(::ArrayStyle, ::Type{T}) where {N,T<:AbstractArray{<:Any,N}} = N
+argdims(::ArrayStyle, ::Type{T}) where {N,T<:LogicalIndex{<:Any,<:AbstractArray{Bool,N}}} = N
+@generated function argdims(s::ArrayStyle, ::Type{T}) where {N,T<:Tuple{Vararg{<:Any,N}}}
     e = Expr(:tuple)
     for p in T.parameters
         push!(e.args, :(ArrayInterface.argdims(s, $p)))
     end
     Expr(:block, Expr(:meta, :inline), e)
 end
 
+"""
+    UnsafeIndex(::ArrayStyle, ::Type{I})
+
+`UnsafeIndex` controls how indices that have been bounds checked and converted to
+native axes' indices are used to return the stored values of an array. For example,
+if the indices at each dimension are single integers than `UnsafeIndex(inds)` returns
+`UnsafeGetElement()`. Conversely, if any of the indices are vectors then `UnsafeGetCollection()`
+is returned, indicating that a new array needs to be reconstructed. This method permits
+customizing the terminal behavior of the indexing pipeline based on arguments passed
+to `ArrayInterface.getindex`. New subtypes of `UnsafeIndex` should define `promote_rule`.
+"""
+abstract type UnsafeIndex end
+
+struct UnsafeGetElement <: UnsafeIndex end
+
+struct UnsafeGetCollection <: UnsafeIndex end
+
+UnsafeIndex(x, i) = UnsafeIndex(x, typeof(i))
+UnsafeIndex(x, ::Type{I}) where {I} = UnsafeIndex(ArrayStyle(x), I)
+UnsafeIndex(s::ArrayStyle, i) = UnsafeIndex(s, typeof(i))
+UnsafeIndex(::ArrayStyle, ::Type{I}) where {I} = UnsafeGetElement()
+UnsafeIndex(::ArrayStyle, ::Type{I}) where {I<:AbstractArray} = UnsafeGetCollection()
+
+Base.promote_rule(::Type{X}, ::Type{Y}) where {X<:UnsafeIndex,Y<:UnsafeGetElement} = X
+
+@generated function UnsafeIndex(s::ArrayStyle, ::Type{T}) where {N,T<:Tuple{Vararg{<:Any,N}}}
+    if N === 0
+        return UnsafeGetElement()
+    elseif N === 1
+        return UnsafeIndex(s, T.parameters[1])
+    else
+        out = typeof(UnsafeIndex(s, T.parameters[1]))
+        for i in 2:N
+            out = promote_type(out, typeof(UnsafeIndex(s, T.parameters[i])))
+        end
+        return out()
+    end
+end
```

Comment:
If users are likely to add their own `UnsafeIndex` methods, being `@generated` is likely to cause problems if people want to define `UnsafeIndex` for their own `Type{T}`. But maybe that isn't a big deal, since defining it for their own Style is probably more likely?

Reply:
> But maybe that isn't a big deal, since defining it for their own Style is probably more likely?

That' what I was originally assuming but it would be nice if defining a new instance of `UnsafeIndex` didn't require redefining that method too.